Activities for Http Connections, Dns, Sockets and SslStream

antonfirsov opened the pull request:
*Edit: for the final design see:*

https://github.com/dotnet/runtime/pull/103922#issuecomment-2221811114
https://github.com/open-telemetry/semantic-conventions/pull/1192

----

This is a replacement for #101814 following our agreement that Http Connection sub-operations (Dns, Socket, SslStream) should be represented with their sub-activities instead of emitting events on the Http Connection activity.

The PR implements the following activity-tree across multiple libraries:

```
System.Net.Http.Connections.HttpConnection
  |- System.Net.NameResolution.DsnLookup
  |- System.Net.Sockets.Connect
  |- System.Net.Security.TlsHandshake
```

Where `System.Net.Http.Connections.HttpConnection` doesn't have a parent (has it's own Trace ID), and every request is linked to their connection via `activity.AddLink()` when the connection is being dispatched to a request.

The 4 new activities have their own ActivitySources which have to be enabled separately. By using `OpenTelemetry.Exporter.Console` this can be done via wildcards:

```csharp
using var tracerProvider = Sdk.CreateTracerProviderBuilder()
                .SetResourceBuilder(ResourceBuilder.CreateDefault().AddService("ConnAct"))
                .AddSource("System.Net.*")
                .AddConsoleExporter()
                .Build();
```

[A sampe program & output can be found here.](https://gist.github.com/antonfirsov/b990472b739a7bbac42f66eb1929b9fd).

Fixes #93832.

/cc @samsp-msft @noahfalk @lmolkova

antonfirsov committed:
Activities for Http Connection, Dns, Sockets and Tls

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/HttpConnectionPool.cs:
```diff
@@ -561,31 +562,37 @@ public async ValueTask<HttpResponseMessage> SendWithVersionDetectionAndRetryAsyn
             IPEndPoint? remoteEndPoint = null;
+
+            // Connection activities should be new roots and not parented under whatever
+            // request happens to be in progress when the connection is started.
```
Why is this? What goes wrong if we don't do this?

antonfirsov replied:
The connection attempt spawned by a request may not be the connection serving the request, thus creating a parent-child relationship would be unfortunate.

Instead, we decided to use activity links to link the connection that is actually serving the request at the moment a connection is associated to a request from the pool.

noahfalk replied:
Also, even if we ignore or change that implementation detail:

For requests, N requests share 1 connection. However Activities are a tree, 1 connection activity can not have N distinct parents.

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/HttpConnectionPool.cs:
```diff
@@ -670,9 +677,10 @@ private async ValueTask<Stream> ConnectToTcpHostAsync(string host, int port, Htt
                         }
                         else
                         {
+                            IPAddress[] addresses = Dns.GetHostAddresses(host);
```
Was the implementation already doing this exact same call elsewhere, or is this allocation new?

stephentoub commented on src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:
```diff
@@ -651,9 +651,15 @@ private static bool LogFailure(object hostNameOrAddress, long? startingTimestamp
+            NameResolutionActivity activity = NameResolutionTelemetry.Log.BeforeResolution(key);
+            if (NameResolutionActivity.IsTracingEnabled())
+            {
+                // Do not overwrite Activity.Current in the caller's ExecutionContext.
```
?

stephentoub replied:
> so we cannot disable tracing

I'm not following. The IsTracingEnabled method is doing:
```
s_activitySource.HasListeners()
```
What prevents the last listener from unregistering before this if block?

antonfirsov replied:
I misunderstood you and thought you want me to avoid the manual reset of `Activity.Current`.

> What prevents the last listener from unregistering before this if block?

Yes, that's a problem, will fix it.

tarekgh replied:
Usually trace collectors are interested in the Activity `Stop` event. So, if the listeners stopped listening before Activity `Stop` event, the whole tracing in the `Activity.Current` will be lost as no-one interested in collecting. 

> Unfortunately there is no way to start a new activity without overwriting Activity.Current.
> shouldn't this be a supported scenario?

I am not sure I understand the scenario here, if you start a new `Activity`, the newly started activity will be set to `Activity.Current`. When you stop it, the original  `Activity.Current` will be restored again. What exactly do you want to achieve here?

CC @noahfalk

antonfirsov replied:
@tarekgh the value of `Activity.Current` depends on the `ExecutionContext` since it's being stored in an `AsyncLocal`. Here we are starting the activity in the parent context (sync path) before kicking off the child Task where the activity is being stopped. This means Start and Stop will manipulate different AsyncLocals.

tarekgh replied:
Thanks. 

I was wondering about your statement `Unfortunately there is no way to start a new activity without overwriting Activity.Current.`. So, I was wondering what is your scenario that needs to start activity and not setting it to the Activity.Current?

antonfirsov replied:
Maybe it's more clear with code:

```C#
public static Task MyAsyncApi()
{
    Activity? activity = new Activity("MyAsyncApi").Start();

    DoSomeSyncStuff();

    return Task.Run(Impl);

    void Impl()
    {
        DoBlockingIO();
        activity?.Stop();
    }
}


Task task = MyAsyncApi();
Console.WriteLine(Activity.Current.OperationName); // prints MyAsyncApi

await task;
Console.WriteLine(Activity.Current.OperationName); // still prints MyAsyncApi :(((
```

The only way to ensure that `Activity.Current` will not be set to `MyAsyncApi` on the parent context forever is to reset `Activity.Current` after starting the `MyAsyncApi` Activity.

noahfalk replied:
I'm not sure what improvements we could realisticaly make? Its an invariant of the AsyncLocal design that the value doesn't change across an async call. I agree with you that it feels awkward/asymmetric that thread-local values flow out of synchronous calls but async-local values do not flow out of async calls, but its hard for me to imagine that design would change in the future.

For now I think the guidance is:
- Preferably start and stop your activities within the same frame. I'd guess 95+% of usage lands here.
- Failing that, be aware of the restrictions in AsyncLocal flow and account for them as needed

stephentoub commented on src/libraries/System.Net.Security/src/System/Net/Security/SslStream.IO.cs:
```diff
@@ -106,15 +111,23 @@ private Task ProcessAuthenticationAsync(bool isAsync = false, CancellationToken
-                return isAsync ?
-                    ForceAuthenticationAsync<AsyncReadWriteAdapter>(IsServer, null, cancellationToken) :
-                    ForceAuthenticationAsync<SyncReadWriteAdapter>(IsServer, null, cancellationToken);
+                activity?.Stop();
```
Won't this stop the activity possibly before authentication has completed? There's no await above.

antonfirsov committed:
address review feedback

tarekgh commented on src/libraries/System.Net.Security/src/System/Net/Security/SslStream.IO.cs:
```diff
@@ -111,30 +111,32 @@ private Task ProcessAuthenticationAsync(bool isAsync = false, CancellationToken
+                startingTimestamp = 0;
+            }
+
+            Activity? activity = s_activitySource.StartActivity(ActivityName, IsServer ? ActivityKind.Server : ActivityKind.Client);
```
nit: you don't need the `finally` block. you can do `using Activity? activity = s_activitySource.StartActivity(ActivityName, IsServer ? ActivityKind.Server : ActivityKind.Client);` which will automatically stop the activity when exiting the block.

davidfowl commented:
>A sampe program & output can be found here..

Can I ask you to show an example with the aspire dashboard? https://aspiredashboard.com/

It should be a simple docker command to run, pointing the app at it with the otlp exporter. Looking at console output for traces is quite hard.

antonfirsov committed:
Merge branch 'main' into connection-activities-05

# Conflicts:
#	src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs
#	src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionBase.cs

antonfirsov committed:
resolve conflicts

antonfirsov committed:
get rid of finally block

antonfirsov commented:
@davidfowl here you go:

![image](https://github.com/dotnet/runtime/assets/6835152/b927bc08-7e18-406f-b1ef-0af60bd80ff6)

Where expanding the connection will see:

![image](https://github.com/dotnet/runtime/assets/6835152/bea5ca77-b01d-4f9d-83ad-3440e692a0d4)

A few notes:

- The connection span covers the entire connection lifetime until closure, while the subspans cover the networking operations (DNS lookup, TCP connection establishment, TLS handshake)
- There are no attributes since thy are not yet implemented. We are currently discussing what is reasonable to include.
- Currently, it is not possible to navigate to connections from requests. For that, Aspire dashboard needs to implement https://github.com/dotnet/aspire/issues/2577.
![image](https://github.com/dotnet/runtime/assets/6835152/f2ef067d-9cc9-4fa9-a30b-89daeafdf812)
- With the current design, you will not be able to navigate to failing connections from their initiating requests. For that we would need to add another Activity Link and tag those links with names ("initiated connection" vs "serving connection"). This might raise the concern that we bake too many connection pool implementation details into the telemetry we produce.

stephentoub commented:
> The connection span covers the entire connection lifetime until closure

Is this going to be useful in a trace? I don't know one way or the other, but it seems like it could result in a lot of noise in a trace.

Is there an equivalent span that's only about setting up the connection, from the time it was requested until the time it became usable?

davidfowl commented:
> Is this going to be useful in a trace? I don't know one way or the other, but it seems like it could result in a lot of noise.

> Is there an equivalent span that's only about setting up the connection, from the time it was requested until the time it became usable?

I agree, we care about how long it took to connect, not how long the connection itself lasts. That feels like it could be another span, but not the most useful one. I think about this as zooming into why a request may have taken so long.

antonfirsov commented:
> That feels like it could be another span, but not the most useful one.

The proposed connection span matches the semantics of the [`http.client.request.duration`](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/dotnet/dotnet-http-metrics.md#metric-httpclientrequestduration) metric.

One solution could be to decompose it as following:

```
[http] connection {peer}                 // Covers the entire pooled connection lifetime
  |- [http] connection_initiation {peer} // Covers the connection establishment lifetime
      |- DNS {host}
      |- [socket] connection_initiation {peer}
      |- TLS {server_name}
```

Another is to only introduce `connection_initiation`, however that feels problematic to link from follow-up requests reusing the same connection.

cc @noahfalk @lmolkova

lmolkova commented:
The overall connection span could be useful to understand:
- which connection my request was done over (if we can link connection span to request span). alternatives could be:
  - some connection id attribute on the span
  - just logs
- could help debugging abandoned response streams and connections that are not returned to the pool (some additional stuff is probably needed to make it obvious)
- not sure if HTTP pipelining is still a thing, but make it easy to detect
- how connection ended (server closed, error happened, etc), when did it end comparing to the request 

Many other useful things are already exposed as metrics (number of active connections, connection duration), so I agree it's usefulness is limited.

Regarding the noise: connections should be rare comparing to requests (if there is any significant load). The spans are reported by a different `ActivitySource`, so users decide if they want them.

Anyway, I don't have a strong opinion on how necessary it is - it can always be added later.

samsp-msft commented:
> > Is this going to be useful in a trace? I don't know one way or the other, but it seems like it could result in a lot of noise.
> 
> > Is there an equivalent span that's only about setting up the connection, from the time it was requested until the time it became usable?
> 
> I agree, we care about how long it took to connect, not how long the connection itself lasts. That feels like it could be another span, but not the most useful one. I think about this as zooming into why a request may have taken so long.

One possibility would be to make the span about connection creation rather than its entire lifetime. It also means that it will be written quicker - spans are only emitted once they have finished.
The reason for wanting to track the connection establishment is so you can figure out why it's taking too long. Is it DNS, sockets, or TLS - and you'll get that data from the span durations.

One reason for knowing about the connection lifetime is if it suddenly starts changing - and being shorter. You could get that from a metric rather than a trace. The extra data on a trace isn't going to tell you much.

antonfirsov commented:
> One possibility would be to make the span about connection creation

The challange here is that if we have only a `connection_initiation` span, we would need to link that from *all* requests that reuse the same connection, even ones which come in way after the `connection_initiation` span's end time (otherwise we would just link an arbitrary request that landed on the connection first - also weird). Wouldn't this be confusing?

davidfowl commented:
>The challange here is that if we have only a connection_initiation span, we would need to link that from all requests that reuse the same connection, even ones which come in way after the connection_initiation span's end time. Wouldn't this be confusing?

You mean all of the ones that are waiting on the same connection to be established right? Not all requests using an already established connection.

antonfirsov commented:
> You mean all of the ones that are waiting on the same connection to be established right? Not all requests using an already established connection.

There is no way to link multiple requests to a connection being established. There is a trivial way to link an already established connection when it has been dispatched from the pool to a request. There is a possibility to create a _"this connection has been initiated by this request, but it might be served by a different connection"_ link.

Consequentially, our options are:
1. Link the `connection_initiation` span to all requests served by the _established_ connection, even if the request comes way after connection initiation.
2. Link the `connnection_initiation` span to the request that initiated the connection. Cons: it might be not the connection serving the request.
3. Link the `connection_initiation` span to only to it's initiating request after the connection has been dispatched from the pool. Cons: the higher the concurrency, the lower the likehood of these links.

I assumed we are going with option 1, since 2-3 seemed to be even less logical.

We can also have multiple links, or even better, we can implement [this](https://github.com/dotnet/runtime/pull/103922#issuecomment-2192784359), making sure the links have meaningful semantics :)

lmolkova commented:
From the offline discussion

```
[http] connection {peer}                 // Covers the entire pooled connection lifetime
  |- [http] connection_initiation {peer} // Covers the connection establishment
      |- DNS {host}
      |- [socket] connection_initiation {peer}
      |- TLS {server_name}
```

`[http] connection_initiation {peer}` feels like an artificial thing that's only necessary to group DNS, socket and TLS. 
Without it, they are still grouped and have individual durations.

BTW, I'm surprised we [don't have metric](https://learn.microsoft.com/en-us/dotnet/core/diagnostics/built-in-metrics-system-net#metric-httpclientactive_requests) for how long establishing a connection tool - this is where it'd be much more useful than as a dedicated span

noahfalk commented:
Maybe everyone else is following the 'initiation' terminology fine, but I'm going to avoid it because every time that word comes up I find myself trying to disambiguate how it is being used. Instead I am using these separate terms which seem clearer to me:

- "Injected Connection" - When a request doesn't have any available connection in the pool, it may inject a new one (ie a call to [CheckForHttp11ConnectionInjection](https://source.dot.net/System.Net.Http/R/1442681d18b69afe.html)).
- "Connection Setup" - The set of steps that occur at the beginning of each connection before it is useful, including DNS lookups, TCP handshake, and TLS handshake.


In terms of what to do, I'd suggest something like this...

For each connection create spans as Anton suggested:
```
[http] connection {peer}                 // Entire connection lifetime
  |- [http] connection_setup {peer}      // Covers the connection setup only
      |- DNS {host}
      |- [socket] TCP handshake {peer}
      |- TLS {server_name}
```

To avoid the noise of the top-level connection span we can leave it disabled-by-default in the Aspire telemetry setup. The spans are controlled by separate ActivitySources so it is easy to turn them on and off independently.

I also think we should consider some adjustments to request span (which is hopefully easy to implement)
1. If a request does not immediately acquire a connection from the connection pool ([TryGetPooledHttp11Connection](https://source.dot.net/System.Net.Http/System/Net/Http/SocketsHttpHandler/ConnectionPool/HttpConnectionPool.Http1.cs.html#ac90f83a3cafcc32) returns false) then add a child span called something like "WaitForConnection". The duration of this child is the duration that the request waits to get an available connection from the pool.
2. The "WaitForConnection" span should have a link to the "Connection Setup" span from the serving connection (not the Injected Connection).
3. If the connection lifetime span was created (if users opted to turn it on), then add a link from the primary request span to connection lifetime span.

Here is how I imagine that would play out in some different scenarios:

### Typical request that didn't have to wait for a connection

Request span is identical to what you see today, no children, no added links.
```
<---------------- Request ----------------------->
```

### Request that is slow because it had to wait for connection setup

```
<---------------- Request --------------------------->
<--------- WaitForConnection --------->
```
Following the link in WaitForConnection Span:
```
<--- Linked Connection Setup --------->
<--- DNS --->
             <--- TCP -->
                         <----TLS----->
```

### Request that is slow because connection pool was overloaded

```
                                  <---------------- Request --------------------------->
                                  <--------- WaitForConnection --------->
```
Following the link in WaitForConnection Span:
```
<--- Linked Connection Setup --------->
<--- DNS --->
             <--- TCP -->
                         <----TLS----->
```
The Connection Setup span ending at time T1 and WaitForConnection ending at some much later time T2 indicates that it wasn't connection setup that was causing the delay, the connection was busy handling other requests.


How does that sound?

antonfirsov commented:
> Instead I am using these separate terms which seem clearer to me

I like the proposed terminology! 💯

> 2. The "WaitForConnection" span should have a link to the "Connection Setup" span from the serving connection (not the Injected Connection).

Isn't this redundant given that the `request -> connection -> connection_setup` link?

> The Connection Setup span ending at time T1 and WaitForConnection ending at some much later time T2 indicates that it wasn't connection setup that was causing the delay, the connection was busy handling other requests.

I'm afraid this can be a common case under high load & slow connections, in which case IMO such a link gives less insight than a link to `Injected Connection`:
- `Injected Connection` is always close in time to the request span
- When `Wait For Connection` times out (common issue when there are problems), there will be no `Serving Connection` we could link.

noahfalk commented:
> I'm afraid this can be a common case under high load & slow connections, in which case IMO a Serving Connection link gives less insight than a link to Injected Connection:
>
> - Injected Connection is always close in time to the request span
>  - When a WaitForConnection span times out (common issue when there are problems), there will be no Serving Connection we could link, while I believe we can usually link an Injected Connection.

In the case that wait for connection times out, the request returns to the caller with some exception and is never transmitted right? I'm guessing it means we are in one of these situations:
1. Some requests are being made successfully, but the rate new requests are being queued exceeds the rate that connections can transmit them. We'd probably be operating at the connection limit so most requests would not inject a connection.
2. Connection setup is failing so no suitable connections for this request are ever getting populated in the pool.

For (1) I assume there isn't usually an injected connection to link to and even if there was it doesn't add much useful diagnostic info because the problem they need to diagnose relates to request volume and request handling latency.
For (2) the injected connnection does seem potentially useful because it acts as an example of the connection failures which are preventing the pool from becoming populated.

What if we linked to the injected connection only under the conditions that (a) wait for connection timed out and (b) the injected connection failed setup? If that isn't easy to implement or easy to agree on a convention for it I also think we could just ignore injected connection links entirely for now.

antonfirsov committed:
Merge branch 'main' into connection-activities-05

antonfirsov committed:
Merge branch 'main' into connection-activities-05

antonfirsov committed:
System.Net.NameResolution: record tags and ActivityStatusCode

antonfirsov committed:
set DNS activity DisplayName

antonfirsov committed:
Socket Activity tags

lmolkova commented on src/libraries/System.Net.Security/src/System/Net/Security/SslStream.IO.cs:
```diff
@@ -131,15 +146,21 @@ private async Task ProcessAuthenticationWithTelemetryAsync(bool isAsync, Cancell
-                NetSecurityTelemetry.Log.HandshakeFailed(IsServer, startingTimestamp, ex.Message);
+                if (startingTimestamp is not 0)
+                {
+                    NetSecurityTelemetry.Log.HandshakeFailed(IsServer, startingTimestamp, ex.Message);
```
should we set activity status and error.type here?

antonfirsov replied:
Yepp!

antonfirsov commented on src/libraries/System.Net.Security/src/System/Net/Security/SslStream.IO.cs:
```diff
@@ -120,8 +125,18 @@ private Task ProcessAuthenticationAsync(bool isAsync = false, CancellationToken
+                startingTimestamp = 0;
+            }
+
+            using Activity? activity = s_activitySource.StartActivity(ActivityName, IsServer ? ActivityKind.Server : ActivityKind.Client);
```
Actually, I'm still not convinced this should be Client in case of server auth.

> the server auth is initiated by the server

We can view any server operation as being initiated by the client and the server together. Eg. to establish an incoming TCP request from a client socket, the server socket needs to `accept()` it. Even though the first packet is sent by the client, the the server needs to be actively participating in the handshake. I don't see how/why would a TLS handshake fundamentally differ.

Also, according to the [spec](https://opentelemetry.io/docs/specs/otel/trace/api/#spankind) `Client` spans need a `Server` span on the other side, so we either keep TLS `Client`-`Server`, or we make them both `Internal`. I was wondering if the spec has any hints if `Internal` is a right choice, but unfortunately it doesn't.

> Internal spans represent operations which do not cross a process boundary. 

Although it's true for `SslStream` since it's just a utility to execute TLS handshakes over other streams, a literal interpretation of this definition means to me that we could also define HTTP spans `Internal`, since it's the underlying transport layer's job to cross/not cross the process boundaries. One can keep all the data flow in-process if they want.

lmolkova replied:
Span kind description is really not so helpful - https://github.com/open-telemetry/opentelemetry-specification/pull/4088

In practice, the `server` kind is used as an indication of external application-level request made against the current application. And the generic server throughput could be calculated as the rate of server spans. When someone renders a service-map, they use server spans to render incoming edges to a node representing a service.

TLS handshake on the server side is a tiny piece of server handling and is not comparable to application-level request processing, you probably never want to see it among incoming requests or as an edge on the service map.

So having additional stages of server call handled as server spans would go against current practice.  
The same argument could be made against CLIENT kind (but everyone abuses it already).

TL;DR: I don't have a good answer, but feel strongly against SERVER. Since there is no guidance/prior art, should we do INTERNAL?

lmolkova replied:
ah, just found some prior art in Java instrumentations https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/netty/netty-4.1/javaagent/src/test/groovy/Netty41ConnectionSpanTest.groovy - they use INTERNAL kind for DNS and socket connect. 

Let me check if there was any prior discussion and how they ended up there.

antonfirsov replied:
> Since there is no guidance/prior art, should we do INTERNAL?

If https://github.com/open-telemetry/opentelemetry-specification/pull/4088 cleans up the definition in a way that would aslo encourage making both connect and accept sides of socket INTERNAL (rather than client-server), it would make sense to me.

samsp-msft replied:
As these are not individual network requests (DNS could be, but its a couple of steps removed) I think internal is OK.

lmolkova replied:
span kinds are defined in such a way that it's hard to say what's client and what's internal. DNS calls don't propagate the context and don't have server spans as children.

we also don't know if there would be any real physical DNS call under that span.

So it's a matter of picking the least controversial option among all other options and sticking with it.

antonfirsov committed:
emit attributes on SslStream Activity

antonfirsov committed:
connection setup activity graph

antonfirsov committed:
emit tags on http connection activities

antonfirsov committed:
Merge branch 'connection-activities-05' of https://github.com/antonfirsov/runtime into connection-activities-05

antonfirsov committed:
fix MacOS failure

antonfirsov committed:
adjustments

antonfirsov commented:
We agreed to implement a design derived from the recommendation in https://github.com/dotnet/runtime/pull/103922#issuecomment-2193842336:

- The activitity tree is being extended as following:

```
HttpRequestOut // existing request
  |- wait_for_connection // has link to connection_setup of the SERVING connection

HTTP connection_setup {peer}      // Covers the connection setup only
  |- DNS lookup {host}
  |- socket connect {peer}
  |- TLS handshake {server_name}
```

- All the new activties have their own ActivitySources so users/middlewares can opt-in selectively. This also applies to `wait_for_connection`. The main reason for that is that is that the UI of certain monitoring tools (eg. AzMon) is not prepared to handle links at scale, so `wait_for_connection` has to be off by default in production scenarios.
- The design is defined in https://github.com/open-telemetry/semantic-conventions/pull/1192, adjustments are still expected, but the fundamental shape should not change.

Open questions:
- Activity source name for `WaitForConnection`. The PR currently goes with `System.Net.Http.ConnectionLink.WaitForConnection`.
- SpanKind for TLS spans. See https://github.com/dotnet/runtime/pull/103922#discussion_r1667217073.
- Inclusion/exclusion of certain attributes, minor naming questions, see https://github.com/open-telemetry/semantic-conventions/pull/1192.

Since the fundaments are in good shape, I marked this ready for review.

/cc @davidfowl @ManickaP @MihaZupan

antonfirsov commented:
Update: 

We had another offline discussion where a concern came up with the design: inability to link back from `connection_setup` to all requests the connection served. Given this, and the uncertainity about how will such links be handled by monitoring tools, we decideded to ship this as an **experimental** feature.

We will signal this by adding the `Exprimental` keyword to the ActivitySource names. The most obvious solution would be to put both `ConnectionSetup` and `WaitForConnection` under `System.Net.Http.Experimental`.

According to @samsp-msft, Aspire plans to integrate this feature regardless, including visualization support on Aspire dashboards.

MihaZupan commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs:
```diff
@@ -74,6 +74,7 @@ public partial class SocketAsyncEventArgs : EventArgs, IDisposable
         private Socket? _currentSocket;
         private bool _userSocket; // if false when performing Connect, _currentSocket should be disposed
         private bool _disposeCalled;
+        internal Activity? _activity;
```
```suggestion
        internal Activity? _connectActivity;
```

MihaZupan replied:
Since there was some discussion around whether we need the field or not, if we expect that having the activity would be rare, you could e.g. stash it in a `ConditionalWeakTable`.

(I don't know that it matters more here than e.g. in HTTP though)

antonfirsov replied:
> having the activity would be rare

In .NET 9 where the feature is experimental, extremely rare.

> I don't know that it matters more here than e.g. in HTTP though

This penalizes all Socket code that uses `SocketAsyncEventArgs` directly, including send/receive path. Contra-argument: perf-critical code, such as Kestrel transport should reuse such `SocketAsyncEventArgs` anyways.

@stephentoub should I look into how to get rid of the field using `ConditionalWeakTable`?

stephentoub replied:
The field on SAEA is there purely to know which Activity to use in post-Connect logging? Using a CWT will mean a dictionary lookup after every connect, even if there is no activity for the SAEA instance. We could lazily initialize the CWT itself if the common case will be no requests in the process ever have an activity.

antonfirsov replied:
> The field on SAEA is there purely to know which Activity to use in post-Connect logging?

Yes.

> We could lazily initialize the CWT itself if the common case will be no requests in the process ever have an activity.

This sounds good enough for 9. Later, maybe we can merge a some status fields, introducing a bit indicating if there is an Activity?

MihaZupan commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:
```diff
@@ -77,22 +81,73 @@ private void AcceptFailed(SocketError error, string? exceptionMessage)
+                {
+                    string peerAddress = ipEndPoint.Address.ToString();
+                    int port = ipEndPoint.Port;
+                    activity.DisplayName = $"socket connect {peerAddress}:{port}";
```
Nit: if `IsAllDataRequested` is expected to be rare, you could save a string allocation by using `{ipEndPoint.Address}` directly here (assuming we aren't already allocating the `ToString` elsewhere).

MihaZupan commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:
```diff
@@ -166,6 +221,32 @@ public void DatagramSent()
+            SocketError.AccessDenied => "access_denied",
+            SocketError.ProtocolType => "protocol_type",
+
+            _ => "_OTHER"
```
Should this return `socketError.ToString()` instead, similar to how we fall back to exception types?

antonfirsov replied:
I think reporting an enum member name without converting it to snake case would go against conventions. I'm trying to catch all SocketErrors that can reasonably occur here, users should not see `_OTHER`.

MihaZupan commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/ConnectionSetupDiagnostics.cs:
```diff
@@ -0,0 +1,86 @@
+            Activity? activity = s_waitForConnectionActivitySource.StartActivity(DiagnosticsHandlerLoggingStrings.WaitForConnectionActivityName);
+            if (activity is not null)
+            {
+                activity.DisplayName = $"wait_for_connection {authority.HostValue}:{authority.Port}";
```
We're also using `"HTTP connection_setup"` `"socket connect"`, `"DNS"`, `"TLS client"` ...
Should this be e.g. `HTTP wait_for_connection` instead?
The naming feels a bit random

antonfirsov replied:
+1 for `HTTP wait_for_connection`. @lmolkova thoughts?

MihaZupan commented on src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHandlerLoggingStrings.cs:
```diff
@@ -9,12 +9,17 @@ namespace System.Net.Http
+        public const string RequestActivityStartName       = RequestActivityName + ".Start";
+        public const string RequestActivityStopName        = RequestActivityName + ".Stop";
+
+        public const string ConnectionNamespace            = "System.Net.Http.Connections";
+        public const string ConnectionSetupActivityName    = ConnectionNamespace + ".ConnectionSetup";
+        public const string WaitForConnectionNamespace     = "System.Net.Http.ConnectionLink";
+        public const string WaitForConnectionActivityName  = WaitForConnectionNamespace + ".WaitForConnection";
```
Do we need two namespaces for this?

Edit: I guess it doesn't matter given https://github.com/dotnet/runtime/pull/103922#issuecomment-2215784470

ManickaP reviewed (commented):
Leaving just these 2 small comments here. Miha beat me with most of the stuff, so +1 on his feedback.

antonfirsov commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/ConnectionSetupDiagnostics.cs:
```diff
@@ -0,0 +1,86 @@
+
+namespace System.Net.Http
+{
+    internal static class ConnectionSetupDiagnostics
```
In other libs we start the activity together with recording an `EventSource` or emitting a metric.
Since there is no such strong correlation between HTTP connection telemetry pillars & unlike `HttpTelemetry` this class doesn't have to be built against WASM, separation makes more sense to me.

antonfirsov replied:
PS: note that metrics are also handled in separate classes. In HTTP there is a lot of telemetry logic in total, IMO it would be bad to unify it in a single monster class.

antonfirsov replied:
Agree with the renaming, disagree with code organization preferences 😆

antonfirsov commented on src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionTelemetry.cs:
```diff
@@ -131,5 +138,104 @@ private static string GetHostnameFromStateObject(object hostNameOrAddress)
+                    {
+                        string[]? answerValues = answer switch
+                        {
+                            string h => [h],
```
@lmolkova is there a way to define types like `variant(string | string[])` in semconv?

antonfirsov commented on src/libraries/System.Net.Security/src/System/Net/Security/SslStream.IO.cs:
```diff
@@ -120,9 +125,29 @@ private Task ProcessAuthenticationAsync(bool isAsync = false, CancellationToken
+            using Activity? activity = s_activitySource.StartActivity(ActivityName, IsServer ? ActivityKind.Server : ActivityKind.Client);
+            if (activity is not null)
+            {
+                activity.DisplayName = IsServer ? "TLS server" : $"TLS client {TargetHostName}";
+                if (activity.IsAllDataRequested && !IsServer)
+                {
+                    activity.SetTag("tls.client.server_name", TargetHostName);
+                }
```
This will be renamed to `server.address`, see https://github.com/open-telemetry/semantic-conventions/pull/1192#discussion_r1667406120.

> I assume we could add a tag when we receive one from the client

This is being set for the client only.

antonfirsov committed:
Merge branch 'main' into connection-activities-05

antonfirsov committed:
consolidate connection setup Activity error handling into a single catch block

antonfirsov commented on src/libraries/System.Net.Http/tests/FunctionalTests/DiagnosticsTests.cs:
```diff
@@ -1001,7 +1198,7 @@ public async Task SendAsync_ActivityIsCreatedIfRequested(bool currentActivitySet
                     ActivitySource.AddActivityListener(new ActivityListener
                     {
-                        ShouldListenTo = _ => true,
+                        ShouldListenTo = s => s.Name is "System.Net.Http" or "",
```
Looks like it's not needed at this line but some cases will fail, if the default ActivitySource is not included in L1230. I couldn't figure out why though, there are so many possible execution paths in this test 😅

antonfirsov committed:
address more review feedback

antonfirsov committed:
merge stuff

antonfirsov committed:
emit the link on the request span instead of wait_for_connection

antonfirsov commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs:
```diff
@@ -91,7 +91,7 @@ public Http3Connection(HttpConnectionPool pool, HttpAuthority authority, bool in
         public void InitQuicConnection(QuicConnection connection)
         {
-            MarkConnectionAsEstablished(connection.RemoteEndPoint);
+            MarkConnectionAsEstablished(connectionSetupActivity: null, remoteEndPoint: connection.RemoteEndPoint);
```
https://github.com/dotnet/runtime/pull/104641 is not going as smooth as I wanted, I will handle this as the last item in my queue of things to polish out on this PR.

antonfirsov committed:
Merge branch 'main' into connection-activities-05

# Conflicts:
#	src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs
#	src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHelper.cs
#	src/libraries/System.Net.Http/tests/FunctionalTests/DiagnosticsTests.cs

antonfirsov committed:
adjust System.Net.NameResolution

antonfirsov committed:
adjust System.Net.Sockets naming, add UDS test

antonfirsov committed:
replace the _connectActivity field with a CWT

antonfirsov committed:
Adjust SslStream tracing

antonfirsov committed:
ActivityKind

antonfirsov committed:
adjust HTTP Activities

Commit:
Merge branch 'main' into connection-activities-05

antonfirsov committed:
SocketsHttpHandler_DiagnosticsTest_Http3

antonfirsov committed:
WIP h3

antonfirsov committed:
adjust to main

antonfirsov committed:
implement H/3 connection diagnostics

Commit:
Merge branch 'main' into connection-activities-05

antonfirsov committed:
actually use the same ActivitySource for ConnectionSetup & WaitForConnection

antonfirsov commented:
This is ready for the final round of reviews. Changes since last review:
- Address feedback from latest review.
- Resolve merge conflicts from #104251 and other telemetry-related PRs
- Link `ConnectionSetup` from the `HttpRequestOut` instead of `WaitForConnection`. Make sure the link is present even when there was no wait.
- Prefix every new ActivitySource name and Activity OperationName with `.Experimental` upon agreement with @noahfalk. Full list of HTTP-related Activities:

```
System.Net.Http.HttpRequestOut
Experimental.System.Net.Http.Connections.ConnectionSetup
Experimental.System.Net.Http.Connections.WaitForConnection
Experimental.System.Net.NameResolution.DnsLookup
Experimental.System.Net.Sockets.Connect
Experimental.System.Net.Security.TlsHandshake
```
- Change the Activity kind for all new activities to INTERNAL upon agreement with @lmolkova
- Adjust Activity DisplayNames and attribute names to the current spec. Left a [review](https://github.com/open-telemetry/semantic-conventions/pull/1192#pullrequestreview-2169318070) for the stuff that needs to be changed there instead of the code.
- Added HTTP/3 support for `ConnectionSetup` and `WaitForConnection`

cc @ManickaP @MihaZupan @lmolkova @noahfalk

antonfirsov commented:
Here is a demo how this works now together with #104251:

https://github.com/dotnet/runtime/assets/6835152/8cf7a78a-950a-4deb-b597-3a7c0035adbf

lmolkova commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs:
```diff
@@ -77,22 +81,72 @@ private void AcceptFailed(SocketError error, string? exceptionMessage)
+                        activity.SetTag("network.peer.address", ipEndPoint.Address.ToString());
+                        activity.SetTag("network.peer.port", port);
+                        activity.SetTag("network.type", ipEndPoint.AddressFamily == AddressFamily.InterNetwork ? "ipv4" : "ipv6");
+                    }
```
wdyt about https://github.com/open-telemetry/semantic-conventions/pull/1192#discussion_r1673448739 ? 
adding `network.transport` if it's `udp` or `unix`

ManickaP replied:
This would also make sense for QUIC in the future.

antonfirsov replied:
UDS doesn't map well to the OSI semantics used by the [`network.*`](https://opentelemetry.io/docs/specs/semconv/attributes-registry/network/) attributes. If anything, it's rather in the network than in the transport layer: sockets are created with `AddressFamily=Unix`, `ProtocolType=Unspecified`.

From .NET perspective "QUIC is not sockets": we should not emit Activities from the same `System.Net.Sockets` activity span. Also, mapping QUIC to OSI would mean that `network.type=ipv4/6` and `network.transport=udp`, which is useless from diagnostics perspective.

IMO semconv has [some homework](https://github.com/open-telemetry/semantic-conventions/issues/1237) to do here, until then I prefer to avoid emitting stuff based on assumptions. My prefrence would be to migrate away from the archaic OSI semantics, it's unfortunate that the related `network.*` attributes are marked as stable already.

lmolkova replied:
semconv tell to set `network.transport=unix` 

https://github.com/open-telemetry/semantic-conventions/blob/f17a4bf973d32a0888a8219a408397f8eeac12ed/docs/attributes-registry/network.md?plain=1#L96

I agree UDS does not map well to OSI, so whatever we did would be imperfect, but this ship has sailed, it's stable

antonfirsov replied:
Sorry, missed that one! QUIC is still unclear though :)

lmolkova replied:
there seem to be is no controversy in adding quic as one of the values of `network.transport` - I replied on the semconv issue. It fits into the OSI levels and the attribute definition perfectly. 
The `network.transport` is an enum and all enums in semconv are open - so it should be fine to add it without it being mentioned explicitly.

> `network.transport` has the following list of well-known values. If one of them applies, then the respective value MUST be used; otherwise, a custom value MAY be used

I can send the PR to semconv to add it, but it'd probably takes a few days to merge.

MihaZupan commented on src/libraries/System.Net.Security/src/System/Net/Security/SslStream.IO.cs:
```diff
@@ -120,9 +125,29 @@ private Task ProcessAuthenticationAsync(bool isAsync = false, CancellationToken
+            using Activity? activity = s_activitySource.StartActivity(ActivityName);
+            if (activity is not null)
+            {
+                activity.DisplayName = IsServer ? "TLS server" : $"TLS client {TargetHostName}";
```
Should this include "handshake"?
E.g. From your video, other activities have the action in the display name (connect, wait for connection, ...)

antonfirsov replied:
[Agree](https://github.com/open-telemetry/semantic-conventions/pull/1192#discussion_r1673902817). Will change it unless there is pushback.

lmolkova replied:
no pushback :)

MihaZupan reviewed (approved):
I'll let Mana double check the H3 part, but this LGTM, thank you

P.S. When we eventually remove the experimental prefix, my small preference would be to use the existing activity source for HTTP.

ManickaP commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/ConnectionSetupDiagnostics.cs:
```diff
@@ -0,0 +1,98 @@
+
+        public static Activity? StartWaitForConnectionActivity(HttpAuthority authority)
+        {
+            Activity? activity = s_connectionsActivitySource.StartActivity(DiagnosticsHandlerLoggingStrings.WaitForConnectionActivityName);
```
This is starting the activity unconditionally, is that what we want (no HasListeners check)?
Moreover, for H/3 this is even called without check for any diagnostics being on (that is part of `HttpConnectionWaiter` line 22).

antonfirsov replied:
No, this will return `null` cheaply, if there are no listeners. The `HasListeners()` check is only beneficial if we want to do some custom logic depending on whether tracing is on before actually starting an activity.

ManickaP reviewed (commented):
HTTP for now, I have skipped tests though.
I'm continuing on DNS/Sockets/TLS.

ManickaP commented on src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionTelemetry.cs:
```diff
@@ -131,5 +138,103 @@ private static string GetHostnameFromStateObject(object hostNameOrAddress)
+    {
+        private const string ActivitySourceName = "Experimental.System.Net.NameResolution";
+        private const string ActivityName = ActivitySourceName + ".DnsLookup";
+        private static readonly ActivitySource s_activitySource = new(ActivitySourceName);
```
NIT: my pet peeve, omitting the class with new makes looking for ctor calls impossible (as "find all references" is borked due to ref sources).
```suggestion
        private static readonly ActivitySource s_activitySource = new ActivitySource(ActivitySourceName);
```

ManickaP commented on src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:
```diff
@@ -16,20 +16,20 @@ public static class Dns
         public static string GetHostName()
         {
-            long startingTimestamp = NameResolutionTelemetry.Log.BeforeResolution(string.Empty);
+            NameResolutionActivity activity = NameResolutionTelemetry.Log.BeforeResolution(string.Empty);
```
If I'm reading this right, turning on any diagnostics will also turn on the experimental Activity creation. Is that what we want?

ManickaP commented on src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:
```diff
@@ -695,7 +704,7 @@ private static Task<TResult> RunAsync<TResult>(Func<object, long, TResult> func,
                             ((ICollection<KeyValuePair<object, Task>>)s_tasks).Remove(new KeyValuePair<object, Task>(key!, task));
                             // Since it was canceled, func(..) had not executed and call AfterResolution it needs to be called here.
-                            NameResolutionTelemetry.Log.AfterResolution(key!, startingTimestamp, new OperationCanceledException());
+                            NameResolutionTelemetry.Log.AfterResolution(key!, activity, new OperationCanceledException());
```
Stopping the activity will set the captured `previousActivity` as the current. This will not accidentally mess anything up? I'm asking because we're starting/stopping the activities here in different tasks / threads.

antonfirsov replied:
Since it's a different (child) task, it should not alter the caller `ExecutionContext`.

ManickaP reviewed (commented):
Some more comments.

antonfirsov commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/ConnectionSetupDiagnostics.cs:
```diff
@@ -0,0 +1,98 @@
+
+        public static void AddConnectionLinkToRequestActivity(Activity connectionSetupActivity)
+        {
+            Debug.Assert(connectionSetupActivity is not null);
```
This pattern seems to be generally around for our telemetry everywhere:

```csharp
if (HttpTelemetry.Log.IsEnabled()) HttpTelemetry.Log.SomeMethod();
```

I assumed this exists to save a few cycles by avoiding calling a method and deepening the stack on the happy path.

antonfirsov commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/HttpConnectionPool.Http1.cs:
```diff
@@ -294,14 +294,14 @@ private async Task InjectNewHttp11ConnectionAsync(RequestQueue<HttpConnection>.Q
         }
 
-        private async ValueTask<HttpConnection> ConstructHttp11ConnectionAsync(bool async, Stream stream, TransportContext? transportContext, HttpRequestMessage request, IPEndPoint? remoteEndPoint, CancellationToken cancellationToken)
+        private async ValueTask<HttpConnection> ConstructHttp11ConnectionAsync(bool async, Stream stream, TransportContext? transportContext, HttpRequestMessage request, Activity? activity, IPEndPoint? remoteEndPoint, CancellationToken cancellationToken)
```
At this point the activity is already stopped, and `activity.Stop()` rolls back `Activity.Current` to the previous one. Even if it worked differently, I would prefer to pass it around whenever possible, since `Activity.Current` accessor is expensive, and async locals are obscure.

antonfirsov commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/ConnectionSetupDiagnostics.cs:
```diff
@@ -0,0 +1,98 @@
+
+        public static void StopConnectionSetupActivity(Activity activity, Exception? exception, IPEndPoint? remoteEndPoint)
+        {
+            Debug.Assert(activity is not null);
```
See previous comment. IMO if we want to get rid of this pattern because we prefer simplicity/readability over saving a few cycles, we should do it for all telemetry not only distributed tracing.

antonfirsov replied:
cc @stephentoub

antonfirsov replied:
Actually, I'm being inconsistent since the Start() methods also would need the check here. I will wait for more feedback before changing anything about this.

ManickaP replied:
`ReportError` has the check inside and not at the call site, that's one that is bothering me (or rather Stop and AddConnectionLink). With Start, the check and code is more complicated so it makes sense to keep it inside.

antonfirsov replied:
`ReportError` is already on an expensive Exception path, so no point bothering with micro-optimizations.
I leave this for now as-is, but I believe we should revisit our telemetry coding patterns/guidelines.

My preference would be clarity over microoptimizations everywhere, but some may disagree.

antonfirsov commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/HttpConnectionPool.cs:
```diff
@@ -565,77 +565,96 @@ public async ValueTask<HttpResponseMessage> SendWithVersionDetectionAndRetryAsyn
-                        stream = await ConnectHelper.EstablishSslConnectionAsync(_sslOptionsProxy, request, async, stream, cancellationToken).ConfigureAwait(false);
-                    }
-                    break;
+            Activity? activity = ConnectionSetupDiagnostics.StartConnectionSetupActivity(IsSecure, OriginAuthority);
```
Good catch, `ws/wss` may report `http/https` wrongly in both Metrics and Distributed Tracing. *If* we want to fix this, I would prefer to do it by opting out from creating an activity for those for now, and cover other protocols later.

IMHO in practice this will unlikely hurt anyone for an experimental feature, so my preference would be to create a tracking issue for .NET 10, where we could fix it for both Metrics and Tracing.

ManickaP replied:
>  it as-is for now, and create a tracking issue for .NET 10

:+1:

antonfirsov committed:
*handshake

antonfirsov committed:
Update src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionTelemetry.cs

Co-authored-by: Marie Píchová <11718369+ManickaP@users.noreply.github.com>

antonfirsov committed:
Merge branch 'connection-activities-05' of https://github.com/antonfirsov/runtime into connection-activities-05

antonfirsov commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/HttpConnectionPool.Http3.cs:
```diff
@@ -271,6 +285,9 @@ private async Task InjectNewHttp3ConnectionAsync(RequestQueue<Http3Connection?>.
                     CreateConnectTimeoutException(oce) :
                     e;
 
+                Debug.Assert(connectionSetupActivity?.IsStopped is not true);
```
In that case we would stop the Activity after calling `connection.InitQuicConnection()` / `MarkConnectionAsEstablished`. This might be harmless, but would be inconsistent with H/1.1 H/2 behavior, which I want to avoid.

antonfirsov replied:
Note that we call `StopConnectionSetupActivity()` at L274. It does more than just stopping the activity (not prepared for double-stopping) and I prefer to do the stop on the success path before the `InitQuicConnection() -> MarkConnetionAsEstablished()` calls in order to be consistent with H/1.1 H/2. I wanted to avoid defining a special method in `ConnectionSetupDiagnostics` that would check for `Activity.IsStopped` just for the H3 case.

I guess this logic deserves more comments?

ManickaP replied:
I see, yeah, if you could mention that the stopping should be done before the other diagnostics work is done in `MarkConnetionAsEstablished` (via `InitQuicConnection`) to be consistent with H/1.1 H/2.

antonfirsov commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/HttpConnectionPool.Http3.cs:
```diff
@@ -88,7 +88,16 @@ internal sealed partial class HttpConnectionPool
                 if (!TryGetPooledHttp3Connection(request, out Http3Connection? connection, out HttpConnectionWaiter<Http3Connection?>? http3ConnectionWaiter))
                 {
-                    connection = await http3ConnectionWaiter.WaitWithCancellationAsync(cancellationToken).ConfigureAwait(false);
+                    using Activity? waitForConnectionActivity = ConnectionSetupDiagnostics.StartWaitForConnectionActivity(authority);
```
Note that we call `StopConnectionSetupActivity()` at L274. It does more than just stopping the activity (not prepared for double-stopping) and I prefer to do the stop on the success path before the `InitQuicConnection() -> MarkConnetionAsEstablished()` calls in order to be consistent with H/1.1 H/2. I wanted to avoid defining a special method in `ConnectionSetupDiagnostics` that would check for `Activity.IsStopped` just for the H3 case.

I guess this logic deserves more comments?

antonfirsov replied:
> the start and stop of this activity should closely follow queueStartingTimestamp.

Implemented.

The problem with the existing logic is that we will emit a very short `wait_for_connection` span and near-zero `http.client.request.time_in_queue` duration metric when there is a connection available in the pool. This is not something I would touch in this PR, but maybe we can improve this in long term?

ManickaP replied:
Agreed, we already keep stream count in H3Connection, we could use it to know if we're gonna wait or not. But in general, it would be good to make the stream opening as close as possible to H2 to avoid these quirks. But that's 10.0 issue. I'll file it.

ManickaP replied:
Filed: https://github.com/dotnet/runtime/issues/104783

antonfirsov committed:
fix H3 logic based on feedback

antonfirsov committed:
ConnectionSetupDiagnostics -> ConnectionSetupDistributedTracing

Commit:
Merge branch 'main' into connection-activities-05

antonfirsov commented:
@ManickaP addressed:

https://github.com/dotnet/runtime/pull/103922#discussion_r1674341249
https://github.com/dotnet/runtime/pull/103922#discussion_r1674349583
class renaming from: https://github.com/dotnet/runtime/pull/103922#discussion_r1674125985

ManickaP commented on src/libraries/System.Net.Sockets/tests/FunctionalTests/System.Net.Sockets.Tests.csproj:
```diff
@@ -78,6 +78,8 @@
     <Compile Include="$(CommonTestPath)System\Net\Sockets\SocketImplementationType.cs"
              Link="SocketCommon\SocketImplementationType.cs" />
     <!-- Common test files -->
+    <Compile Include="$(CommonTestPath)System\Net\ActivityRecorder.cs"
+                 Link="Common\System\Net\ActivityRecorder.cs" />
```
```suggestion
    <Compile Include="$(CommonTestPath)System\Net\ActivityRecorder.cs"
             Link="Common\System\Net\ActivityRecorder.cs" />
```

ManickaP reviewed (approved):
Last batch of comments. Thank you for your patience!

antonfirsov committed:
suggestion

Co-authored-by: Marie Píchová <11718369+ManickaP@users.noreply.github.com>

antonfirsov committed:
suggestion

Co-authored-by: Marie Píchová <11718369+ManickaP@users.noreply.github.com>

antonfirsov committed:
suggestion

Co-authored-by: Marie Píchová <11718369+ManickaP@users.noreply.github.com>

antonfirsov commented on src/libraries/System.Net.Security/src/System/Net/Security/SslStream.IO.cs:
```diff
@@ -131,17 +156,60 @@ private async Task ProcessAuthenticationWithTelemetryAsync(bool isAsync, Cancell
+                }
+            }
+
+            static (string?, string?) GetNameAndVersionString(SslProtocols protocol) => protocol switch
```
EventSource telemetry does the same after acquiring the enum with `GetSslProtocolInternal()` so I assumed it should be a single value after handshake completion normally. If not, our telemetry doesn't emit the protocol info.

https://github.com/dotnet/runtime/blob/f9eda079502aa737ed7c75cf14dea10223cfb0fa/src/libraries/System.Net.Security/src/System/Net/Security/NetSecurityTelemetry.cs#L178-L201

antonfirsov commented on src/libraries/System.Net.Security/src/System/Net/Security/SslStream.IO.cs:
```diff
@@ -15,6 +15,11 @@ namespace System.Net.Security
 {
     public partial class SslStream
     {
+        private const string ActivitySourceName = "Experimental.System.Net.Security";
```
These pieces of code evolved like dust balls in a dirty room, but there is some logic:
- In HTTP, various telemetry pillars follow a complex requirements across multiple protocols and various aspects (request vs connection telemetry, Metrics vs EventSource vs Activities, http1-2 vs http3). When there is cohesion, a class has been defined to collect related things into a single place, when there isn't, things were kept separate.
- In DNS and Sockets, telemetry can be tracked together across the 3 pillars
- In SSL, the Activity logic is so tiny, that I didn't bother to move it to a separate class.

I will do a second round of thinking about the last point.

antonfirsov replied:
Moved around things a bit for SslStream

antonfirsov committed:
Merge branch 'main' into connection-activities-05

antonfirsov committed:
implement 'network.transport'

antonfirsov committed:
SslStream: move Activity management code to NetSecurityTelemetry

antonfirsov committed:
readd assertion and add comment

Commit:
Merge branch 'main' into connection-activities-05

antonfirsov commented:
/azp run runtime-libraries-coreclr outerloop

Commit:
Merge branch 'main' into connection-activities-05

davidfowl commented on src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHandlerLoggingStrings.cs:
```diff
@@ -9,12 +9,16 @@ namespace System.Net.Http
+        public const string RequestActivityStartName       = RequestActivityName + ".Start";
+        public const string RequestActivityStopName        = RequestActivityName + ".Stop";
+
+        public const string ConnectionsNamespace           = "Experimental.System.Net.Http.Connections";
```
Experimental?

antonfirsov replied:
It turned out that our solution results is an unexpected volume of span links for tools like Azure Monitor, blowing up their UI, see https://github.com/open-telemetry/opentelemetry-specification/discussions/4130.

We need to work with OTel and Azure Monitor to introduce a standard way for categorizing links so visualization tools can distinguish them. The plan is to drive  this work forward in order to stabilize the feature in .NET 10.
